### ISSUE TITLE
Error on Vercel Nextjs deploy

### ISSUE THREAD

**Issue by cenguidanos** [nrwl/nx#4032]
## Current Behavior
Deploy Nextjs monorepo gets an error

## Steps to Reproduce

1. Create nx workspace with Nextjs preset.
2. Go to Vercel account.
3. Import repository.
4. Modify default commands on Vercel:

Build command: `nx build <app>`
Directory: `dist/apps/<app>`

### Failure Logs
```bash

Error: A "routes-manifest.json" couldn't be found. This is normally caused by a misconfiguration in your project.
Please check the following, and reach out to support if you cannot resolve the problem:
1. If present, be sure your `build` script in "package.json" calls `next build`.  2. Navigate to your project's settings in the Vercel dashboard, and verify that the "Build Command" is not overridden, or that it calls `next build`.  3. Navigate to your project's settings in the Vercel dashboard, and verify that the "Output Directory" is not overridden. Note that `next export` does **not** require you change this setting, even if you customize the `next export` output directory. Learn More: https://err.sh/vercel/vercel/now-next-routes-manifest


```

### Environment
@nrwl/cli: 10.3.2
@nrwl/next: 10.3.2
next: 10.0.1
typescript: 4.0.3

**Comment by mandarini** [comment]
Hi there @cenguidanos !

Thanks for filing an issue! 

You can solve this issue by choosing "Framework preset: Other"

<img width="761" alt="Screenshot 2020-11-06 at 1 22 06 PM" src="https://user-images.githubusercontent.com/6603745/98360803-29b3c400-2033-11eb-9769-46359b1ea7e1.png">

Take a look at [this issue and this comment here](https://github.com/nrwl/nx/issues/3051#issuecomment-718202172).

Let me know if you need any more assistance, or I can close this issue!

**Comment by cenguidanos** [comment]
Maybe this help works with only SSG mode, but if you use SSR or lamba pages, you get this:

![Image](https://academiares-recursos-web.s3.eu-west-3.amazonaws.com/img/2020-11-06+(2).png)

Thanks

**Comment by mandarini** [comment]
Hi @cenguidanos , this is what I see as well when deploying the next.js app on Vercel. So, I assume that since you see the same screen, that it works, and deployed correctly?

**Comment by cenguidanos** [comment]
Hi @mandarini , thanks for your response but correctly correctly... It's deployed but the webpage don't works and we see a system folder. It's not the point of this. Maybe i don't explain my self very well with english. Sorry. Our interest is deploy a @nrwl/next app in Vercel like normal next.js repo with all opimizations, it's possible for now with nx? Thanks.

**Comment by mandarini** [comment]
Hi there @cenguidanos ! Sorry, you're right, your request makes sense! I'll take a look on this next week! In the meantime, I would suggest you can try some different output directories or configurations on Vercel?

**Comment by cenguidanos** [comment]
On local, or other nodejs platforms, we can do this:

#### 1. Add new script to `package.json`:

```json
{
"scripts": {
    "nx": "nx",
    "next": "next", <---
    "start": "nx serve",
    "build": "nx build",
    "test": "nx test",
    "lint": "nx workspace-lint && nx lint",
    "e2e": "nx e2e",
    "affected:apps": "nx affected:apps",
    "affected:libs": "nx affected:libs",
    "affected:build": "nx affected:build",
    "affected:e2e": "nx affected:e2e",
    "affected:test": "nx affected:test",
    "affected:lint": "nx affected:lint",
    "affected:dep-graph": "nx affected:dep-graph",
    "affected": "nx affected",
    "format": "nx format:write",
    "format:write": "nx format:write",
    "format:check": "nx format:check",
    "update": "nx migrate latest",
    "workspace-schematic": "nx workspace-schematic",
    "dep-graph": "nx dep-graph",
    "help": "nx help"
  },
}
```

#### 2. Build app with `nx build <app>`.
#### 3. Serve app with next `yarn next dist/apps/<app>`.

And works fine:

```bash
yarn run v1.22.5
$ next start dist/apps/<app>
ready - started server on http://localhost:3000
```

Anyway, this don't works with Vercel Hosting, but yes with docker, for example.

**Comment by gregkonush** [comment]
I can confirm this issue as well, my standalone nextjs app was deployed successfully to the vercel, but when I converted it to be part of monorepo setup of nx then builds were failing and deployed correctly

**Comment by cenguidanos** [comment]
Vercel Support Twitter send me this: https://twitter.com/vercel_support/status/1325770048934465537?s=20

```bash
Hello Cristian,

We don't support NX-powered apps at the moment. Our team is working on making this feature available for you shortly, and we should get back to you once we support NX.
```

**Comment by gregkonush** [comment]
I wonder if you can customize NX builder to customize the Next commands that NX is running under the hood.

**Comment by cenguidanos** [comment]
I don't know how to do it. Now we are in Render.com with a simple nodejs server, but is not serverless...

**Comment by gregkonush** [comment]
@cenguidanos I figured the way it works. So for root of application settings you choose the app folder like `apps/dashboard` and then you leave nextjs as framework option and all other things leave as default. Then I created separate `package.json` file specifically for that in its app's folder(apps/dashboard)  that has this:
```
"scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start"
  },
```

**Comment by cenguidanos** [comment]
From Vercel team:

https://twitter.com/paulogdm/status/1326849234176516102?s=19

**Comment by cenguidanos** [comment]
Ok, the solution is change the `workspace.json` and modify *outputPath* to:

```json
{
  "version": 1,
  "projects": {
    "home": {
      "root": "apps/<app>",
      "sourceRoot": "apps/<app>",
      "projectType": "application",
      "schematics": {},
      "architect": {
        "build": {
          "builder": "@nrwl/next:build",
          "options": {
            "root": "apps/home",
            "outputPath": "." <--
          }
```

And apply Next.js Vercel builder on deploy process.

![example](https://academiares-recursos-web.s3.eu-west-3.amazonaws.com/img/code_example.png)

**Comment by gregkonush** [comment]
I am not sure why this closed. I followed the steps that you provided and I don't think that works?

**Comment by cenguidanos** [comment]
You can't? An alternative to modify `workspace.json` is create a build command on `package.json`: `"build:app": "nx build app --outputPath .` and override build command on Vercel. This works for me. Leo Lamprecht, dev on Vercel, says that they are working with optimizarions for NX workspaces. But, for now, he says that this is the solution.

**Comment by JamesHenry** [comment]
In my view this is the most optimal Nx + Next.js + Vercel setup that is available today: https://github.com/nrwl/nx/issues/3051#issuecomment-737482131

**Comment by thesobercoder** [comment]
I'm still getting this error. 😭

**Comment by JamesHenry** [comment]
Hi folks 👋 

We have worked together with Vercel to improve the configuration required to deploy Next.js apps on Vercel from Nx workspaces. Additional/custom scripts are no longer required.

Please see the updated guidance here https://github.com/nrwl/nx/pull/4483, it is available for you to implement immediately.

Once the PR is merged, the updated guidance will also be available on https://nx.dev/latest/react/guides/nextjs

Thanks!

**Comment by cenguidanos** [comment]
Thanks!!

**Comment by cenguidanos** [comment]
Maybe i found a bug on dev mode. 

I add this script in `next.config.js`:

```js
/* eslint-disable @typescript-eslint/no-var-requires */
const withNx = require('@nrwl/next/plugins/with-nx');
const { appendFileSync } = require('fs');

module.exports = withNx({
    webpack: (config, { dev, isServer }) => {
        // Create file only in client development build
        if (dev && !isServer) appendFileSync('data.txt', 'hi dev');

        // Create file only in client production build
        if (!dev && !isServer) appendFileSync('data.txt', 'hi prod');

        return config;
    }
});

```

But this script is being executed with: `nx serve <app>` and `nx  build <app>`. In two modes, development and production. And it's executed [twice](https://nextjs.org/docs/api-reference/next.config.js/custom-webpack-config). The results of the print in file is:

```bash
hi prodhi prod
```

---

next; v10.0.1
react: v17.0.1
nrwl/*: 11.1.1

---

The steps to reproduce is:

1. Create new workspace with next app
2. Select SCSS
3. Add lines in next.config.js

**Comment by cenguidanos** [comment]
> Maybe i found a bug on dev mode.
> 
> I add this script in `next.config.js`:
> 
> ```js
> /* eslint-disable @typescript-eslint/no-var-requires */
> const withNx = require('@nrwl/next/plugins/with-nx');
> const { appendFileSync } = require('fs');
> 
> module.exports = withNx({
>     webpack: (config, { dev, isServer }) => {
>         // Create file only in client development build
>         if (dev && !isServer) appendFileSync('data.txt', 'hi dev');
> 
>         // Create file only in client production build
>         if (!dev && !isServer) appendFileSync('data.txt', 'hi prod');
> 
>         return config;
>     }
> });
> ```
> 
> But this script is being executed with: `nx serve <app>` and `nx build <app>`. In two modes, development and production. And it's executed [twice](https://nextjs.org/docs/api-reference/next.config.js/custom-webpack-config). The results of the print in file is:
> 
> ```shell
> hi prodhi prod
> ```
> 
> next; v10.0.1
> react: v17.0.1
> nrwl/*: 11.1.1
> 
> The steps to reproduce is:
> 
> 1. Create new workspace with next app
> 2. Select SCSS
> 3. Add lines in next.config.js

Works fine **without** plugin:

```js
/* eslint-disable @typescript-eslint/no-var-requires */
const { appendFileSync } = require('fs');

module.exports = {
    target: 'experimental-serverless-trace',
    webpack: (config, { dev, isServer }) => {
        // Create file only in client development build
        if (dev && !isServer) appendFileSync('data.txt', 'hi dev');

        // Create file only in client production build
        if (!dev && !isServer) appendFileSync('data.txt', 'hi prod');

        return config;
    }
};
```

`nx serve <app>`: `hi dev`

**Comment by JamesHenry** [comment]
@cenguidanos Let's try and avoid this being a catch-all issue for Next.js related concerns, please open a new issue with specific context if needed.

In your particular example, though, you may wish to try setting the `NODE_ENV` environment variable.

E.g.

```sh
NODE_ENV=production npx nx build foo --prod
```

I always set `NODE_ENV=production` for my Next.js projects, as in my case I need it for proper purging of unused tailwind classes, but it could be applicable to this as well

**Comment by cenguidanos** [comment]
ok

**Comment by JamesHenry** [comment]
Sorry for the confusion @cenguidanos I didn’t mean for you to close this one - this one is regarding deployments on Vercel, that’s important and until my docs PR is merged it doesn’t have an “official” solution.

I was merely saying that issues with local development and Next.js are out of scope for this specific issue

**Comment by cenguidanos** [comment]
Ok, only for  the case, this is working:

```json
{
  "start": "cross-env NODE_ENV=development nx serve",
  "build": "cross-env NODE_ENV=production nx build",
  "export": "cross-env NODE_ENV=production nx export"
}
```

```js
/* eslint-disable @typescript-eslint/no-var-requires */
const withNx = require('@nrwl/next/plugins/with-nx');

module.exports = withNx({
	webpack: (config) => {
          // Webpack variables
          const path = config.context;
  
          const isServer = config.name === 'server';
          const isClient = config.name === 'client';
  
          const isDev = process.env.NODE_ENV === 'development';
          const isProd = process.env.NODE_ENV === 'production';
          
	  const isDevServer = isDev && isServer;
  
	  if (isDevServer) somethingCool();
  
	  return config;
	}
});

```

**Comment by vsavkin** [comment]
Merged @JamesHenry 's PR with the docs changes (https://github.com/nrwl/nx/pull/4483).

We will deploy it to nx.dev later this week.

I'm going to close this issue, folks. 

Thank you, everyone!

**Comment by gregorybellencontre** [comment]
Hello
I'm trying to deploy a new Nx project into Vercel, but I have the same issue.
The `development` branch deployment works fine, but when I click on "Promote to production", I have the following error:
```
Error! The file "/vercel/path0/dist/apps/web/.next/routes-manifest.json" couldn't be found. This is normally caused by a misconfiguration in your project.
```

It seems to be related to build cache.
There's another thread about this specific behavior : https://github.com/nrwl/nx/issues/8294

**Comment by nalnir** [comment]
Yeah, having the same problem. When building for production it fails. 

My current setup.
<img width="907" alt="Screenshot 2022-09-12 at 08 28 14" src="https://user-images.githubusercontent.com/28774583/189555846-263d93ee-444a-4c62-bdf5-d3ff503decaa.png">

I tried the option where all the fields are set to default. I tried the one with no-cloud flag, but so far it fails exactly at the same point.

**Comment by maapteh** [comment]
The error is a little weird, i also got the routes manifest error but in the end i forgot to add /.next to output dir. The NX guideline is pretty clear now 

![docs](https://nx.dev/documentation/shared/guides/next-deploy-vercel-1.png)

https://nx.dev/recipe/deploy-nextjs-to-vercel
